hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -677,6 +677,7 @@ else if ( maybeProxy instanceof PersistentAttributeInterceptable ) {
 			final PersistentAttributeInterceptable interceptable = (PersistentAttributeInterceptable) maybeProxy;
 			final PersistentAttributeInterceptor interceptor = interceptable.$$_hibernate_getInterceptor();
 			if ( interceptor instanceof EnhancementAsProxyLazinessInterceptor ) {
+				( (EnhancementAsProxyLazinessInterceptor) interceptor ).setSession( session );
```

Comment:
Copying the [conversation from the original PR](https://github.com/hibernate/hibernate-orm/pull/4339#discussion_r742615080).

From @yrodiere:

> I wonder why we have to do that... I see we already have code whose purpose is to set the session after reassociation, e.g. in `org.hibernate.persister.entity.AbstractEntityPersister#afterReassociate`. But for some reason that code isn't called here...

From @busitech:

> `Persist` and `merge` have not been exposing proxy objects to the application until now, and this helped some failing test cases to pass.  One example is CascadeDeleteCollectionTest-->testDetachedOriginal:149 failing with LazyInitializationException during cascade before delete.

Reply:
Alright. I think the "ideal" solution would be to remove some of this code and somehow call `org.hibernate.persister.entity.AbstractEntityPersister#afterReassociate` instead, but I suppose that can wait.